### ISSUE TITLE
faster `copyFile` via zero-copy API's on osx/bsd (`copyfile`), linux (`splice`)

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/RFCs#330]
## proposal 1
* use C `copyfile` on osx and bsd
* implementation: https://github.com/nim-lang/Nim/pull/16883

## proposal 2
* use C `splice` on linux
refs:
  * https://en.wikipedia.org/wiki/Splice_(system_call)
  * https://stackoverflow.com/questions/7463689/most-efficient-way-to-copy-a-file-in-linux

## proposal 3
* on other OS (besides osx, bsd, linux), use the existing copyFile implementation but with raw IO via `open/read/write` instead of libc buffering as done currently to avoid overhead since we can read whole buffers directly (refs https://stackoverflow.com/a/7463735/1426932)

## proposal 4
* use `8192` instead of `8000` for buffer size, or even gradually increase buffer size from 4K to 16K for large files

## proposal 5: improve `copyFileWithPermissions`
* remove race condition and simplify `copyFileWithPermissions` on osx/bsd by calling `copyfile` with `COPYFILE_STAT`

## links
* https://stackoverflow.com/questions/7463689/most-efficient-way-to-copy-a-file-in-linux

**Comment by Araq** [comment]
Not that I mind these ideas, but the stdlib doesn't claim to provide the *fastest* (or "best") way of copying files around. And an implementation should strive for small code size and maintainability.

**Comment by juancarlospaco** [comment]
So far the diff is tiny, and most lines are tests anyway.
:)

**Comment by timotheecour** [comment]
> but the stdlib doesn't claim to provide the fastest (or "best") way of copying files around

I think the stance should be:
* stdlib doesn't claim to provide the fastest/best way to do X, but PR's should always be welcome to improve X on platform Y so long the improvement justifies the added complexity. Relevant criterions to assess this:
* is the improvement (speedup, memory use, added flexibility, etc) meaningful?
* maintenance cost (are there tests in place? is it tested in CI?)
* does it add new (especially non standard) dependencies?
* does it make it harder to evolve the API?

> And an implementation should strive for small code size and maintainability.

all else being equal, yes, but not at the expense of performance where it matters.
Small (generated asm) code size is what the compiler `--lean` is for (ping on https://github.com/nim-lang/Nim/pull/14282 which introduces this generally useful flag), and in future work can be used in more places to provide leaner implementations in places where this matters most.

**Comment by rominf** [comment]
I'm proposing to change 2nd proposal: I suggest using `sendfile` instead of `splice`, as it's more specific and easier to use, and also `sendfile` is implemented using `splice`, see:
https://stackoverflow.com/a/4483342/2108548
https://code.woboq.org/linux/linux/fs/read_write.c.html#do_sendfile

We can find a readable example of `sendfile` usage in stdlib of Julia: https://github.com/JuliaLang/julia/blob/6468dcb04ea2947f43a11f556da9a5588de512a0/base/filesystem.jl#L116

What do you think?

**Comment by timotheecour** [comment]
> I suggest using sendfile instead of splice

https://stackoverflow.com/a/7464280/1426932
> Unfortunately, you cannot use sendfile() here because the destination is not a socket. (The name sendfile() comes from send() + "file").

**Comment by rominf** [comment]
In Linux kernels before 2.6.33, out_fd must refer to a socket. Since Linux 2.6.33 it can be any file. If it is a regular file, then sendfile() changes the file offset appropriately.
See: https://man7.org/linux/man-pages/man2/sendfile.2.html

PS: StackOverflow is not the absolute source of Truth :-)

**Comment by rominf** [comment]
And yes, I would use `sendfile`, as even Debian oldstable comes with a newer version of kernel.

**Comment by timotheecour** [comment]
Yes, you are correct, I should've double checked ; just added a comment in https://stackoverflow.com/questions/7463689/most-efficient-way-to-copy-a-file-in-linux/7464280#comment117018927_7464280

**Comment by timotheecour** [comment]
@rominf you'll also need to check for interrupted writes and restart from where it left off (maxes out at 2GB IIRC) along with the usual EINTR etc.

**Comment by rominf** [comment]
Is it OK to rewrite this code: https://github.com/python/cpython/blob/1b57426e3a7842b4e6f9fc13ffb657c78e5443d4/Lib/shutil.py#L114

**Comment by timotheecour** [comment]
check also boot::filesystem, it recently added `sendfile` support to speedup their `copy_file` API: refs https://github.com/boostorg/filesystem/commit/9182b4caa34f14a246f3bcd6cae5ad9cb270682d
(but look at latest sources instead of that commit)

note also this code:
```nim

#if defined(BOOST_FILESYSTEM_USE_SENDFILE)
    // sendfile started accepting file descriptors as the target in Linux 2.6.33
    if (major > 2u || (major == 2u && (minor > 6u || (minor == 6u && patch >= 33u))))
      cfd = &copy_file_data_sendfile;
#endif

#if defined(BOOST_FILESYSTEM_USE_COPY_FILE_RANGE)
    // Although copy_file_range appeared in Linux 4.5, it did not support cross-filesystem copying until 5.3
    if (major > 5u || (major == 5u && minor >= 3u))
      cfd = &copy_file_data_copy_file_range;
#endif
```

which favors using the more recent `copy_file_range` for recent enough linux kernel version.

## refs 
https://man7.org/linux/man-pages/man2/copy_file_range.2.html

**Comment by FedericoCeratto** [comment]
sendfile is fast and has been around for a long time and it should be used by the stdlib also for sockets and [async]httpserver. Related: #304 https://github.com/nim-lang/Nim/issues/9716#issuecomment-439053314 https://github.com/nim-lang/Nim/issues/4334#issuecomment-225906512
